Signal/src/views/QuotedReplyPreview.swift

```diff
@@ -0,0 +1,120 @@
+//
+//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+
+@objc
+protocol QuotedReplyPreviewDelegate: class {
+    func quotedReplyPreviewDidPressCancel(_ preview: QuotedReplyPreview)
+}
+
+@objc
+class QuotedReplyPreview: UIView {
+    public weak var delegate: QuotedReplyPreviewDelegate?
+
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    init(quotedMessage: TSQuotedMessage) {
+        super.init(frame: .zero)
+
+        let isQuotingSelf = quotedMessage.authorId == TSAccountManager.localNumber()
+
+        // used for stripe and author
+        // FIXME actual colors TBD
+        let authorColor: UIColor = isQuotingSelf ? .ows_materialBlue : .black
+
+        // used for text and cancel
+        let foregroundColor: UIColor = .darkGray
+
+        let authorLabel: UILabel = UILabel()
+        authorLabel.textColor = authorColor
+        if isQuotingSelf {
+            authorLabel.text = NSLocalizedString("MEDIA_GALLERY_SENDER_NAME_YOU", comment: "")
+        } else {
+            authorLabel.text = Environment.current().contactsManager.displayName(forPhoneIdentifier: quotedMessage.authorId)
+        }
+        authorLabel.font = .ows_dynamicTypeHeadline
+
+        let bodyLabel: UILabel = UILabel()
+        bodyLabel.textColor = foregroundColor
+        bodyLabel.font = .ows_footnote
+
+        bodyLabel.text = {
+            if let contentType = quotedMessage.contentType {
+                let emoji = TSAttachmentStream.emoji(forMimeType: contentType)
+                return "\(emoji) \(quotedMessage.body ?? "")"
+            } else {
+                return quotedMessage.body
+            }
+        }()
+
+        let thumbnailView: UIView? = {
+            if let image = quotedMessage.thumbnailImage() {
```

Comment:
See the design in abstract.  I suspect we'll want to make this pretty similar - if not exactly similar - to the way quoted messages are shown in the conversation.

* The "quote stripe" should be on the left of just the text, not the quoted author.
* If there's no thumbnail, we should use a circle with a content-appropriate icon.
* If there's no quoted text/caption, we should show some text.

I propose that I pull out my work (not yet merged) to render quoted content in the conversation views into a separate cell (since it's closer to the current design and has some extra measurement functionality we need in the cells) and we try to reuse it here?

Reply:
I'll add this to this list of things to go over with myles.